frictionless_ckan_mapper/ckan_to_frictionless.py

```diff
@@ -150,14 +150,49 @@ def dataset(ckandict):
     # package_show can have license_id and license_title
     # TODO: do we always license_id i.e. can we have license_title w/o
     # license_id?
-    if ('licenses' not in outdict and 'license_id' in outdict):
-        outdict['licenses'] = [{
-            'name': outdict['license_id'],
-        }]
-        if 'license_title' in outdict:
-            outdict['licenses'][0]['title'] = outdict['license_title']
+    # There's no mention in the docs of license_title nor license_url, plus
+    # license_id is listed as optional in both CKAN 2.8 and 2.9.
+
+    # Looping like this because all those keys are optional.
+    for key in ['license_id', 'license_title', 'license_url']:
+        if key in outdict:
+            outdict['licenses'] = [{}]
+            break  # check to create list of dicts only once
+    if 'license_id' in outdict:
+        outdict['licenses'][0]['name'] = outdict['license_id']
+    if 'license_title' in outdict:
+        outdict['licenses'][0]['title'] = outdict['license_title']
+    if 'license_url' in outdict:
+        outdict['licenses'][0]['path'] = outdict['license_url']
     outdict.pop('license_id', None)
     outdict.pop('license_title', None)
+    outdict.pop('license_url', None)
+
+    # Check `ckandict` instead of `outdict` because outdict['extras']
+    # was deleted earlier
+    licenses_there = False  # First assume we won't find licenses
+    if ckandict.get('extras'):
```

Comment:
For this chunk of code (lines 173 to 195), I think you will find it much easier to follow by referring to the accompanying tests in `tests/test_ckan_to_frictionless.py`:

* `test_dataset_license`
* `test_dataset_license_with_licenses_in_extras`

It's still a bit of a mess, but that's JSON. :shrug:

Reply:
I don't understand this logic quite @sglavoie why do i do this afterward. The normal logic would be: if i have already unpacked extras and had licenses i just use that. Why am i unpacking this again??

Reply:
I agree. The idea was:

* Add everything from 'extras' first. That means also 'licenses' if it's there.
* After that, outdict['extras'] is actually removed from the returned dict (line 102).
* If we find one of 'license_id', 'license_title', 'license_url', then we need to consider this information for the first license that will appear in 'licenses', so it just resets that list to contain an empty dict for now.
* We still have access to everything that was passed as an argument from 'ckandict', so we make sure to rebuild the list of licenses from it because it may have been cleared if one of 'license_id', 'license_title', 'license_url' was there.
* Even if one of those keys wasn't there, it will loop and see if there are licenses and only add them if they weren't there already. In this case, the licenses will be there already if we don't remove the 'licenses' when finding one of  'license_id', 'license_title', 'license_url'.